### PR TITLE
dezony IQ x openEMS - Hackathon 24.02.2023

### PR DESCRIPTION
**Pull request by hoip**
Integration of dezony IQ in openEMS.

dezony communicates with openEMS via REST API. 

Here you can find our [Swagger documentation](https://app.swaggerhub.com/apis/reddo/dezony/1.1.0)

### CODE REVIEW

**Comment by hoip** [comment]
@mlang97 Can you please assign a review to the merge request?

**Comment by mlang97** [comment]
Thank you @hoip for your PR. @sebastianasen and me will have a look at this PR.

File: io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/Config.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	@AttributeDefinition(name = "Alias", description = "Human-readable name of this Component; defaults to Component-ID")
+	String alias() default "";
+
+	@AttributeDefinition(name = "Is enabled?", description = "Is this Component enabled?")
+	boolean enabled() default true;
+
+	@AttributeDefinition(name = "Debug Mode", description = "Activates the debug mode")
+	boolean debugMode() default false;
+
+	@AttributeDefinition(name = "IP-Address", description = "The IP address of the charging station. If the charger has two connectors, the second/slave evcs has the IP 192.168.25.31.", required = true)
```
Comment (sebastianasen) [inline io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/Config.java:23]: Please remove the description hint for the second evcs or adjust the IP to your needs / defaults.

File: io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/Config.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	String ip() default "192.168.50.88";
+	
+	@AttributeDefinition(name = "Port", description = "The IP address of the charging station. If the charger has two connectors, the second/slave evcs has the IP 192.168.25.31.", required = true)
```
Comment (sebastianasen) [inline io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/Config.java:26]: Please adapt the description

File: io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/DezonyApi.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+	            return JsonUtils.parseToJsonObject(body);
+	        }
+	    } catch (IOException e) {
+	        // Log the error
+	        e.printStackTrace();
+	    }
+
+	    // If there was an error, set the communication failed flag to true and return null
+	    this.dezonyImpl._setChargingstationCommunicationFailed(true);
```
Comment (sebastianasen) [inline io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/DezonyApi.java:97]: The value is not reset to false again when the connection is re-established.
Maybe you could do it like in the sendPostRequest method.

**Review by sebastianasen — CHANGES_REQUESTED** [review]
Thank you for implementing the charger!
Find my comments below.
Feel free to let us know how you continue with OpenEMS, or if you have any problems in production in the Community Forum.

File: io.openems.edge.evcs.dezony/.settings/org.eclipse.buildship.core.prefs [hunk 4]
```diff
@@ -0,0 +1,2 @@
+connection.project.dir=..
+eclipse.preferences.version=1
```
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.evcs.dezony/.settings/org.eclipse.buildship.core.prefs:2]: Yes, please revert it to defaults - should work automatically with prepare-commit.sh

File: io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/DezonyImpl.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+@EventTopics({ //
+		EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE, //
+		EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE, //
+})
+public class DezonyImpl extends AbstractManagedEvcsComponent
+		implements OpenemsComponent, EventHandler, Dezony, Evcs, ManagedEvcs {
+	protected Config config;
+	protected DezonyApi api;
+	protected final Logger log = LoggerFactory.getLogger(DezonyImpl.class);
+	private final DezonyReadWorker readWorker = new DezonyReadWorker(this);
```
[the remark this answers is not in the document]
Comment (hoip) [reply io.openems.edge.evcs.dezony/src/io/openems/edge/evcs/dezony/DezonyImpl.java]: I am not quite sure in which order the parameters should be reordered, so I have already followed the production order of another integration.

**Review by sebastianasen — APPROVED** [review]
Thank you for adapting my change requests.
@sfeilmeier Decide if you want to take another look.
I think we could change this to a generic http worker #2100 later.

**Review by sfeilmeier — APPROVED** [review]
I left some comments and ToDos; some I fixed myself, others should be handled in future Pull Requests.

Thank you @hoip, @mlang97 and @sebastianasen, I really appreciate the patience, endurance and professionality that went into this Pull Request.

**Pull request merged**